src/pudl/dfc.py

```diff
@@ -0,0 +1,154 @@
+        if not filename.startswith("gs://"):
+            # Do not make directories when dealing with remote storage.
+            # TODO(rousik): this is fairly crude solution and won't work
+            # for non gcs remote storage.
+            Path(os.path.dirname(filename)).mkdir(exist_ok=True, parents=True)
+        data.to_pickle(filename)
+        self._table_ids[name] = self._instance_id
+
+    def add_reference(self, name: str, table_id: uuid.UUID):
+        """Adds reference to a named dataframe to this collection.
+
+        This assumes that the data is already present on disk.
+        """
+        self._table_ids[name] = table_id
+
+    def __getitem__(self, name: str) -> pd.DataFrame:
+        """Allows accessing dataframes via self[name]."""
+        return self.get(name)
+
+    def __setitem__(self, name: str, data: pd.DataFrame):
+        """Allows adding dataframes via self[name] = value."""
+        return self.store(name, data)
+
+    def items(self) -> Iterator[Tuple[str, pd.DataFrame]]:
+        """Iterates over table names and the corresponding pd.DataFrame objects."""
+        for name in self.get_table_names():
+            yield (name, self.get(name))
+
+    def get_table_names(self) -> List[str]:
+        """Returns sorted list of dataframes that are contained in this collection."""
+        return sorted(set(self._table_ids))
+
+    def get_table_ids(self) -> Dict[str, uuid.UUID]:
+        """Returns dict mapping dataframe names to their uuid identifiers."""
+        return dict(self._table_ids)
+
+    @staticmethod
+    def make_from_dict(d: Dict[str, pd.DataFrame]):
+        """Constructs new DataFrameCollection from dataframe dictionary."""
+        return DataFrameCollection(**d)
+
+    def to_dict(self) -> Dict[str, pd.DataFrame]:
+        """Loads the entire collection to memory as a dictionary."""
+        return dict(self.items())
+
+    def update(self, other):
+        """Adds references to tables from the other DataFrameCollection."""
+        # TODO(rousik): typecheck other?
+        for name, table_id in other.get_table_ids().items():
+            self.add_reference(name, table_id)
+
+    def union(self, *others):
+        """Returns new DataFrameCollection that is union of self and others."""
+        # TODO(rousik): annotation for others: List[DataFrameCollection] does not work due
+        # to python idiosyncracies. There is a way but it is a weird one using generics.
+        dfc = DataFrameCollection()
+        dfc.update(self)
+        for col in others:
+            dfc.update(col)
+        return dfc
+
+
+@task
+def merge(left: DataFrameCollection, right: DataFrameCollection):
+    """Merges two DataFrameCollection instances."""
+    return left.union(right)
+
+
+@task
+def merge_list(list_of_dfc: List[DataFrameCollection]):
+    """Merges list of DataFrameCollection instancs."""
+    return DataFrameCollection().union(*list_of_dfc)
```

Comment:
What does it mean for these to be Prefect `@task`s? Why is that helpful / necessary?